### Importing Libraries

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, classification_report


### Loading the Data 

#### This is the first set of data (_1)

In [2]:
# Importing the 
#This is the first set of split of the data
x_h1n1_train_1 = pd.read_csv('x_h1n1_train_1.csv')
x_h1n1_test_1 = pd.read_csv('x_h1n1_test_1.csv')
y_h1n1_train_1 = pd.read_csv('y_h1n1_train_1.csv')
y_h1n1_test_1 = pd.read_csv('y_h1n1_test_1.csv')


x_seasonal_train_1 = pd.read_csv('x_seasonal_train_1.csv')
x_seasonal_test_1 = pd.read_csv('x_seasonal_test_1.csv')
y_seasonal_train_1 = pd.read_csv('y_seasonal_train_1.csv')
y_seasonal_test_1 = pd.read_csv('y_seasonal_test_1.csv')


#### This is the second set of data (_2)

In [3]:
#This is the second split of the data
x_h1n1_train_2 = pd.read_csv('x_h1n1_train_2.csv')
x_h1n1_test_2 = pd.read_csv('x_h1n1_test_2.csv')
y_h1n1_train_2 = pd.read_csv('y_h1n1_train_2.csv')
y_h1n1_test_2 = pd.read_csv('y_h1n1_test_2.csv')

x_seasonal_train_2 = pd.read_csv('x_seasonal_train_2.csv')
x_seasonal_test_2 = pd.read_csv('x_seasonal_test_2.csv')
y_seasonal_train_2 = pd.read_csv('y_seasonal_train_2.csv')
y_seasonal_test_2 = pd.read_csv('y_seasonal_test_2.csv')


#### Just taking a look at part of the loaded data.

In [4]:
x_seasonal_train_1.head(1)

h1n1_concern_Not at all concerned  h1n1_concern_Not very concerned  \
0                                  0                                0   

   h1n1_concern_Somewhat concerned  h1n1_concern_Unknown  \
0                                1                     0   

   h1n1_concern_Very concerned  h1n1_knowledge_A little knowledge  \
0                            0                                  1   

   h1n1_knowledge_A lot of knowledge  h1n1_knowledge_No knowledge  \
0                                  0                            0   

   h1n1_knowledge_Unknown  \
0                       0   

   behavioral_antiviral_meds_Not on antiviral medications  ...  \
0                                                  1       ...   

   employment_occupation_qxajmpny  employment_occupation_rcertsgn  \
0                               0                               0   

   employment_occupation_tfqavkke  employment_occupation_ukymxvdu  \
0                               0                               0   

   employment_occupation_uqqtjvyb  employment_occupation_vlluhbov  \
0                               0                               0   

   employment_occupation_xgwztkwe  employment_occupation_xqwwgdyp  \
0                               1                               0   

   employment_occupation_xtkaffoo  employment_occupation_xzmlyyjv  
0                               0                               0  

[1 rows x 183 columns]

In [5]:
x_h1n1_train_1.head(1)

h1n1_concern_Not at all concerned  h1n1_concern_Not very concerned  \
0                                  0                                0   

   h1n1_concern_Somewhat concerned  h1n1_concern_Unknown  \
0                                1                     0   

   h1n1_concern_Very concerned  h1n1_knowledge_A little knowledge  \
0                            0                                  1   

   h1n1_knowledge_A lot of knowledge  h1n1_knowledge_No knowledge  \
0                                  0                            0   

   h1n1_knowledge_Unknown  \
0                       0   

   behavioral_antiviral_meds_Not on antiviral medications  ...  \
0                                                  1       ...   

   employment_occupation_qxajmpny  employment_occupation_rcertsgn  \
0                               0                               0   

   employment_occupation_tfqavkke  employment_occupation_ukymxvdu  \
0                               0                               0   

   employment_occupation_uqqtjvyb  employment_occupation_vlluhbov  \
0                               0                               0   

   employment_occupation_xgwztkwe  employment_occupation_xqwwgdyp  \
0                               1                               0   

   employment_occupation_xtkaffoo  employment_occupation_xzmlyyjv  
0                               0                               0  

[1 rows x 183 columns]

In [6]:
x_seasonal_test_1.head(1)

h1n1_concern_Not at all concerned  h1n1_concern_Not very concerned  \
0                                  0                                0   

   h1n1_concern_Somewhat concerned  h1n1_concern_Unknown  \
0                                1                     0   

   h1n1_concern_Very concerned  h1n1_knowledge_A little knowledge  \
0                            0                                  1   

   h1n1_knowledge_A lot of knowledge  h1n1_knowledge_No knowledge  \
0                                  0                            0   

   h1n1_knowledge_Unknown  \
0                       0   

   behavioral_antiviral_meds_Not on antiviral medications  ...  \
0                                                  1       ...   

   employment_occupation_qxajmpny  employment_occupation_rcertsgn  \
0                               0                               0   

   employment_occupation_tfqavkke  employment_occupation_ukymxvdu  \
0                               0                               0   

   employment_occupation_uqqtjvyb  employment_occupation_vlluhbov  \
0                               0                               0   

   employment_occupation_xgwztkwe  employment_occupation_xqwwgdyp  \
0                               0                               0   

   employment_occupation_xtkaffoo  employment_occupation_xzmlyyjv  
0                               0                               0  

[1 rows x 183 columns]

In [7]:
x_h1n1_train_2.head(1)

h1n1_concern_Not at all concerned  h1n1_concern_Not very concerned  \
0                                  0                                0   

   h1n1_concern_Somewhat concerned  h1n1_concern_Unknown  \
0                                1                     0   

   h1n1_concern_Very concerned  h1n1_knowledge_A little knowledge  \
0                            0                                  1   

   h1n1_knowledge_A lot of knowledge  h1n1_knowledge_No knowledge  \
0                                  0                            0   

   h1n1_knowledge_Unknown  \
0                       0   

   behavioral_antiviral_meds_Not on antiviral medications  ...  \
0                                                  1       ...   

   employment_occupation_qxajmpny  employment_occupation_rcertsgn  \
0                               0                               0   

   employment_occupation_tfqavkke  employment_occupation_ukymxvdu  \
0                               0                               0   

   employment_occupation_uqqtjvyb  employment_occupation_vlluhbov  \
0                               0                               0   

   employment_occupation_xgwztkwe  employment_occupation_xqwwgdyp  \
0                               1                               0   

   employment_occupation_xtkaffoo  employment_occupation_xzmlyyjv  
0                               0                               0  

[1 rows x 162 columns]

In [8]:
x_seasonal_train_2.head(1)

behavioral_antiviral_meds_Not on antiviral medications  \
0                                                  1        

   behavioral_antiviral_meds_On antiviral medications  \
0                                                  0    

   behavioral_antiviral_meds_Unknown  behavioral_avoidance_Avoided  \
0                                  0                             1   

   behavioral_avoidance_Did not avoid  behavioral_avoidance_Unknown  \
0                                   0                             0   

   behavioral_face_mask_Did not wear mask  behavioral_face_mask_Unknown  \
0                                       1                             0   

   behavioral_face_mask_Wore mask  behavioral_wash_hands_Did not wash hands  \
0                               0                                         0   

   ...  employment_occupation_qxajmpny  employment_occupation_rcertsgn  \
0  ...                               0                               0   

   employment_occupation_tfqavkke  employment_occupation_ukymxvdu  \
0                               0                               0   

   employment_occupation_uqqtjvyb  employment_occupation_vlluhbov  \
0                               0                               0   

   employment_occupation_xgwztkwe  employment_occupation_xqwwgdyp  \
0                               1                               0   

   employment_occupation_xtkaffoo  employment_occupation_xzmlyyjv  
0                               0                               0  

[1 rows x 153 columns]

### Creating the Logistic Regression Model

In [9]:
logmodel = LogisticRegression()


In [10]:
logmodel.fit(x_h1n1_train_1, y_h1n1_train_1.values.ravel())


c:\Users\Duane\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [11]:
h1n1_predictions_1 = logmodel.predict(x_h1n1_test_1)

In [12]:
classification_report(y_h1n1_test_1, h1n1_predictions_1)

'              precision    recall  f1-score   support\n\nNot received       0.87      0.94      0.90      4172\n    Received       0.71      0.49      0.58      1170\n\n    accuracy                           0.84      5342\n   macro avg       0.79      0.72      0.74      5342\nweighted avg       0.83      0.84      0.83      5342\n'

In [13]:
confusion_matrix(y_h1n1_test_1, h1n1_predictions_1)

array([[3937,  235],
       [ 598,  572]], dtype=int64)

In [14]:
logmodel.fit(x_seasonal_train_1, y_seasonal_train_1.values)

c:\Users\Duane\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Duane\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [15]:
seasonal_peredictions_1 = logmodel.predict(x_seasonal_test_1)

In [16]:
classification_report(y_seasonal_test_1, seasonal_peredictions_1)

'              precision    recall  f1-score   support\n\nNot received       0.79      0.82      0.80      2856\n    Received       0.78      0.75      0.76      2486\n\n    accuracy                           0.79      5342\n   macro avg       0.78      0.78      0.78      5342\nweighted avg       0.78      0.79      0.78      5342\n'

In [17]:
confusion_matrix(y_seasonal_test_1, seasonal_peredictions_1)

array([[2330,  526],
       [ 622, 1864]], dtype=int64)

### Here is the second set of data.

In [18]:
logmodel.fit(x_h1n1_train_2, y_h1n1_train_2.values.ravel())

c:\Users\Duane\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [19]:
h1n1_predictions_2 = logmodel.predict(x_h1n1_test_2)

In [20]:
classification_report(y_h1n1_test_2, h1n1_predictions_2)

'              precision    recall  f1-score   support\n\nNot received       0.86      0.94      0.90      4172\n    Received       0.70      0.47      0.56      1170\n\n    accuracy                           0.84      5342\n   macro avg       0.78      0.71      0.73      5342\nweighted avg       0.83      0.84      0.83      5342\n'

In [21]:
confusion_matrix(y_h1n1_test_2, h1n1_predictions_2)

array([[3937,  235],
       [ 623,  547]], dtype=int64)

In [22]:
logmodel.fit(x_seasonal_train_2, y_seasonal_train_2.values)

c:\Users\Duane\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Duane\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [23]:
x_seasonal_predictions_2 = logmodel.predict(x_seasonal_test_2)

In [24]:
classification_report(y_seasonal_test_2, x_seasonal_predictions_2)

'              precision    recall  f1-score   support\n\nNot received       0.79      0.81      0.80      2856\n    Received       0.78      0.75      0.76      2486\n\n    accuracy                           0.78      5342\n   macro avg       0.78      0.78      0.78      5342\nweighted avg       0.78      0.78      0.78      5342\n'

In [25]:
confusion_matrix(y_seasonal_test_2, x_seasonal_predictions_2)

array([[2322,  534],
       [ 630, 1856]], dtype=int64)

In [36]:
h1n1_predictions_1

array(['Not received', 'Not received', 'Received', ..., 'Not received',
       'Not received', 'Not received'], dtype=object)

In [35]:
seasonal_peredictions_1

array(['Received', 'Received', 'Received', ..., 'Received',
       'Not received', 'Not received'], dtype=object)

### Here I am combining the predictions into one dataFrame so I can compute the Target.

In [45]:
df = pd.concat([ pd.DataFrame(h1n1_predictions_1),pd.DataFrame(seasonal_peredictions_1)], axis=1)

In [46]:
df.columns = ['h1n1_vaccine' if i==0 else 'seasonal_vaccine' if i==1 else col for i, col in enumerate(df.columns)]

In [47]:
df

h1n1_vaccine seasonal_vaccine
0     Not received         Received
1     Not received         Received
2         Received         Received
3     Not received         Received
4     Not received         Received
...            ...              ...
5337  Not received     Not received
5338  Not received         Received
5339  Not received         Received
5340  Not received     Not received
5341  Not received     Not received

[5342 rows x 2 columns]

In [53]:
Received_Vaccines = pd.DataFrame()
Received_Vaccines['vaccine_status'] = np.select(conditions, choices, default='Unknown')
conditions = [
    (df['h1n1_vaccine'] == 'Not received') & (df['seasonal_vaccine'] == 'Not received'),
    (df['h1n1_vaccine'] == 'Received') & (df['seasonal_vaccine'] == 'Not received'),
    (df['h1n1_vaccine'] == 'Not received') & (df['seasonal_vaccine'] =='Received'),
    (df['h1n1_vaccine'] == 'Received') & (df['seasonal_vaccine'] == 'Received')
]

choices = [
    'No Vaccines',
    'H1N1 Vaccine Only',
    'Seasonal Vaccine Only',
    'Both Vaccines'
]


In [54]:
Received_Vaccines

vaccine_status
0     Seasonal Vaccine Only
1     Seasonal Vaccine Only
2             Both Vaccines
3     Seasonal Vaccine Only
4     Seasonal Vaccine Only
...                     ...
5337            No Vaccines
5338  Seasonal Vaccine Only
5339  Seasonal Vaccine Only
5340            No Vaccines
5341            No Vaccines

[5342 rows x 1 columns]

## Summary

In this notebook I have ran the Logistic Regression model on the 2 sets of data from the previous notebook. The frist(_1) set of data is the entire data set. The second (_2) set of data excludes all h1n1 data in the seasonal vaccine results and excludes the seasonal data from the h1n1 results. The outcome is nearly identical with the first set being marginally more accurate. I can go back and do more feature engineering to possibly get a more accuarte results whcih are 83% and 79% respectively. The final targeted results are in the Received_Vaccines dataFrame.